src/pretty.rs

```diff
@@ -17,16 +23,32 @@ fn atoms_to_doc<'a>(i: &mut usize, atoms: &'a [Atom], indent_level: isize) -> Rc
             return doc;
         } else {
             doc = doc.append(match atom {
-                Atom::Blankline => RcDoc::hardline().append(RcDoc::hardline()),
+                Atom::Blankline => RcDoc::hardline()
+                    .append(RcDoc::hardline())
+                    .append(RcDoc::concat(repeat(RcDoc::space()).take(*indent_level))),
                 Atom::Empty => RcDoc::text(""),
-                &Atom::Hardline => RcDoc::hardline(),
-                Atom::Leaf { content, .. } => RcDoc::text(content.trim_end()),
+                &Atom::Hardline => RcDoc::hardline()
+                    .append(RcDoc::concat(repeat(RcDoc::space()).take(*indent_level))),
+                Atom::Leaf {
+                    content,
+                    single_line_no_indent,
+                    ..
+                } => {
+                    if *single_line_no_indent {
+                        RcDoc::hardline().append(RcDoc::text(content.trim_end()))
+                    } else {
+                        RcDoc::text(content.trim_end())
+                    }
+                }
                 Atom::Literal(s) => RcDoc::text(s),
                 Atom::MultilineOnlyLiteral { .. } => unreachable!(),
                 Atom::IndentEnd => unreachable!(),
                 Atom::IndentStart => {
                     *i += 1;
-                    atoms_to_doc(i, atoms, indent_level).nest(indent_level)
+                    *indent_level += indent_offset;
+                    let res = atoms_to_doc(i, atoms, indent_offset, indent_level);
+                    *indent_level -= indent_offset;
```

Comment:
`indent_level` is `usize`, so if the query rules aren't correct, this will panic when `indent_offset > indent_level`. Previously the indentation was signed, which I suppose was done to get around the panicking, despite resulting in the formatter returning clipped output.

TBH, I think panicking is better than weird output. However, maybe it's worth an issue to handle this error more gracefully.

Reply:
By virtue of L48, we know that `indent_level >= indent_offset`...so maybe this never happens. However, it is clearly possible to write a query with unbalanced indentation:

```scm
(#language! ocaml)
(comment) @append_hardline @append_indent_end
```

Formats:

```
(* Line 1 *)
(* Line 2 *)
```

...as:

```
(* Line 1 *)
```

So it doesn't panic :shrug: As such, ignore this comment, because it's not relevant to your changes! Nonetheless, I'll add an issue to capture this more generally.

Reply:
(Added Issue #299)